File: common-system-test/src/main/java/com/teamscale/test/commons/SystemTestUtils.java
package com.teamscale.test.commons;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.apache.commons.lang3.SystemUtils;
import org.conqat.lib.commons.io.ProcessUtils;
import org.jetbrains.annotations.NotNull;

import com.teamscale.report.testwise.model.TestInfo;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.http.Body;
import retrofit2.http.POST;
import retrofit2.http.PUT;

/**
 * Utilities for running system tests.
 */
public class SystemTestUtils {

	/**
	 * The port for the mock Teamscale server that was picked by the Gradle build script and is guaranteed to not
	 * conflict with other system tests.
	 */
	public static final Integer TEAMSCALE_PORT = Integer.getInteger("teamscalePort");

	/**
	 * The port for the agent that was picked by the Gradle build script and is guaranteed to not conflict with other
	 * system tests.
	 */
	public static final Integer AGENT_PORT = Integer.getInteger("agentPort");

	/**
	 * Turns the coverage of the given {@link TestInfo} into a string for simple assertions.
	 * <p>
	 * Example: {@code file1.java:1,7-12;file2.java:9-22,33}
	 */
	public static String getCoverageString(TestInfo info) {
		return info.paths.stream().flatMap(path -> path.getFiles().stream())
				.map(file -> file.fileName + ":" + file.coveredLines).collect(
						Collectors.joining(";"));
	}

	/**
	 * Runs the clean and verify goal of the Maven project at the given path.
	 *
	 * @throws IOException if running Maven fails.
	 */
	public static void runMavenTests(String mavenProjectPath) throws IOException {
		runMavenTests(mavenProjectPath, new String[0]);
	}


	/**
	 * Runs the clean and verify goal of the Maven project at the given path with the provided arguments
	 *
	 * @throws IOException if running Maven fails.
	 */
	public static void runMavenTests(String mavenProjectPath, String... args) throws IOException {
		String[] allArguments = new String[2 + args.length];
		allArguments[0] = "clean";
		allArguments[1] = "verify";
		System.arraycopy(args, 0, allArguments, 2, args.length);
		runMaven(mavenProjectPath, allArguments);
	}

	/**
	 * Runs Maven in the given Maven project path with the given arguments.
	 *
	 * @throws IOException if running Maven fails.
	 */
	public static void runMaven(String mavenProjectPath, String... mavenArguments) throws IOException {
		ProcessUtils.ExecutionResult result;
		try {
			result = ProcessUtils.execute(buildMavenProcess(mavenProjectPath, mavenArguments));
		} catch (IOException e) {
			throw new IOException("Failed to run ./mvnw clean verify in directory " + mavenProjectPath, e);
		}

		// in case the process succeeded, we still log stdout and stderr in case later assertions fail. This helps
		// debug test failures
		System.out.println("Maven stdout: " + result.getStdout());
		System.out.println("Maven stderr: " + result.getStderr());

		if (result.terminatedByTimeoutOrInterruption()) {
			throw new IOException("Running Maven failed: " + result.getStdout() + "\n" + result.getStderr());
		}
	}

	/**
	 * Runs Gradle in the given Gradle project path with the given arguments.
	 *
	 * @throws IOException if running Gradle fails.
	 */
	public static void runGradle(String gradleProjectPath, String... gradleArguments) throws IOException {
		ProcessUtils.ExecutionResult result;
		try {
			result = ProcessUtils.execute(buildGradleProcess(gradleProjectPath, gradleArguments));
		} catch (IOException e) {
			throw new IOException("Failed to run ./gradlew clean verify in directory " + gradleProjectPath, e);
		}

		// in case the process succeeded, we still log stdout and stderr in case later assertions fail. This helps
		// debug test failures
		System.out.println("Gradle stdout: " + result.getStdout());
		System.out.println("Gradle stderr: " + result.getStderr());

		if (result.terminatedByTimeoutOrInterruption()) {
			throw new IOException("Running Gradle failed: " + result.getStdout() + "\n" + result.getStderr());
		}
	}

	/**
	 * Creates the command-line arguments that can be passed to {@link ProcessBuilder} to invoke Maven with the given
	 * arguments.
	 */
	@NotNull
	public static ProcessBuilder buildMavenProcess(String mavenProjectDirectory, String... mavenArguments) {
		List<String> arguments = new ArrayList<>();
		if (SystemUtils.IS_OS_WINDOWS) {
			Collections.addAll(arguments, "cmd", "/c", "mvnw.cmd");
		} else {
			arguments.add("./mvnw");
		}

		arguments.addAll(Arrays.asList(mavenArguments));

		return new ProcessBuilder(arguments).directory(new File(mavenProjectDirectory));
	}

	/**
	 * Creates the command-line arguments that can be passed to {@link ProcessBuilder} to invoke Gradle with the given
	 * arguments.
	 */
	@NotNull
	public static ProcessBuilder buildGradleProcess(String gradleProjectDirectory, String... gradleArguments) {
		List<String> arguments = new ArrayList<>();
		if (SystemUtils.IS_OS_WINDOWS) {
			Collections.addAll(arguments, "cmd", "/c", "gradlew.bat");
		} else {
			arguments.add("./gradlew");
		}

		arguments.addAll(Arrays.asList(gradleArguments));
		
		return new ProcessBuilder(arguments).directory(new File(gradleProjectDirectory));
	}

	/** Retrieve all files in the `tia/reports` folder sorted by name. */
	public static List<Path> getReportFileNames(String mavenProjectPath) throws IOException {
		try (Stream<Path> stream = Files.walk(Paths.get(mavenProjectPath, "target", "tia", "reports"))) {
			return stream.filter(Files::isRegularFile).sorted().collect(Collectors.toList());
		}
	}

	private interface AgentService {
		/** Dumps coverage */
		@POST("/dump")
		Call<Void> dump();

		/** Changes the current partition. */
		@PUT("/partition")
		Call<Void> changePartition(@Body RequestBody newPartition);

		/** Changes the current partition. Convenience method to pass a plain string. */
		default Call<Void> changePartition(String newPartition) {
			return changePartition(RequestBody.create(MediaType.parse("text/plain"), newPartition));
		}
	}

	/** Instructs the agent via HTTP to dump the currently collected coverage. */
	public static void dumpCoverage(int agentPort) throws IOException {
		new Retrofit.Builder().baseUrl("http://localhost:" + agentPort).build()
				.create(AgentService.class).dump().execute();
	}

	/** Instructs the agent via HTTP to change the partition. */
	public static void changePartition(int agentPort, String newPartition) throws IOException {
		new Retrofit.Builder().baseUrl("http://localhost:" + agentPort).build()
				.create(AgentService.class).changePartition(newPartition).execute();
	}

}

File: common-system-test/src/main/kotlin/com/teamscale/test/commons/SystemTestUtils.kt
package com.teamscale.test.commons

import com.teamscale.report.testwise.model.TestInfo
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import org.apache.commons.lang3.SystemUtils
import org.conqat.lib.commons.io.ProcessUtils
import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.create
import retrofit2.http.Body
import retrofit2.http.POST
import retrofit2.http.PUT
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import kotlin.streams.asSequence

/**
 * Utilities for running system tests.
 */
object SystemTestUtils {
	/**
	 * The port for the mock Teamscale server that was picked by the Gradle build script and is guaranteed to not
	 * conflict with other system tests.
	 */
	@JvmField
	val TEAMSCALE_PORT: Int = Integer.getInteger("teamscalePort")

	/**
	 * The port for the agent that was picked by the Gradle build script and is guaranteed to not conflict with other
	 * system tests.
	 */
	@JvmField
	val AGENT_PORT: Int = Integer.getInteger("agentPort")

	/**
	 * Turns the coverage of the given [TestInfo] into a string for simple assertions.
	 *
	 *
	 * Example: `file1.java:1,7-12;file2.java:9-22,33`
	 */
	@JvmStatic
	val TestInfo.coverage
		get() = paths.flatMap { it.files }.joinToString(";") { "${it.fileName}:${it.coveredLines}" }

	/**
	 * Runs the clean and verify goal of the Maven project at the given path with the provided arguments
	 *
	 * @throws IOException if running Maven fails.
	 */
	@JvmStatic
	@Throws(IOException::class)
	fun runMavenTests(mavenProjectPath: String, vararg args: String) {
		val allArguments = mutableListOf("clean", "verify").apply {
			addAll(args)
		}
		runMaven(mavenProjectPath, *allArguments.toTypedArray())
	}

	/**
	 * Runs Maven in the given Maven project path with the given arguments.
	 *
	 * @throws IOException if running Maven fails.
	 */
	@JvmStatic
	@Throws(IOException::class)
	fun runMaven(mavenProjectPath: String, vararg mavenArguments: String) {
		val result: ProcessUtils.ExecutionResult
		try {
			result = ProcessUtils.execute(buildMavenProcess(mavenProjectPath, *mavenArguments))
		} catch (e: IOException) {
			throw IOException("Failed to run ./mvnw clean verify in directory $mavenProjectPath", e)
		}

		// in case the process succeeded, we still log stdout and stderr in case later assertions fail. This helps
		// debug test failures
		println("Maven stdout: ${result.stdout}")
		println("Maven stderr: ${result.stderr}")

		if (result.terminatedByTimeoutOrInterruption()) {
			throw IOException("Running Maven failed: ${result.stdout}\n${result.stderr}")
		}
	}

	/**
	 * Runs Gradle in the given Gradle project path with the given arguments.
	 *
	 * @throws IOException if running Gradle fails.
	 */
	@JvmStatic
	@Throws(IOException::class)
	fun runGradle(gradleProjectPath: String, vararg gradleArguments: String): ProcessUtils.ExecutionResult {
		val result: ProcessUtils.ExecutionResult
		try {
			result = ProcessUtils.execute(buildGradleProcess(gradleProjectPath, *gradleArguments))
		} catch (e: IOException) {
			throw IOException("Failed to run ./gradlew clean verify in directory $gradleProjectPath", e)
		}

		// in case the process succeeded, we still log stdout and stderr in case later assertions fail. This helps
		// debug test failures
		println("Gradle stdout: ${result.stdout}")
		println("Gradle stderr: ${result.stderr}")

		if (result.terminatedByTimeoutOrInterruption()) {
			throw IOException("Running Gradle failed: ${result.stdout}\n${result.stderr}")
		}
		return result
	}

	/**
	 * Creates the command-line arguments that can be passed to [ProcessBuilder] to invoke Maven with the given
	 * arguments.
	 */
	@JvmStatic
	fun buildMavenProcess(mavenProjectDirectory: String, vararg mavenArguments: String): ProcessBuilder {
		val arguments = mutableListOf<String>()
		if (SystemUtils.IS_OS_WINDOWS) {
			arguments.addAll(listOf("cmd", "/c", "mvnw.cmd"))
		} else {
			arguments.add("./mvnw")
		}

		arguments.addAll(listOf(*mavenArguments))

		return ProcessBuilder(arguments).directory(File(mavenProjectDirectory))
	}

	/**
	 * Creates the command-line arguments that can be passed to [ProcessBuilder] to invoke Gradle with the given
	 * arguments.
	 */
	private fun buildGradleProcess(gradleProjectDirectory: String, vararg gradleArguments: String): ProcessBuilder {
		val arguments = mutableListOf<String>()
		if (SystemUtils.IS_OS_WINDOWS) {
			arguments.addAll(listOf("cmd", "/c", "gradlew.bat"))
		} else {
			arguments.add("./gradlew")
		}

		arguments.addAll(listOf(*gradleArguments))

		return ProcessBuilder(arguments).directory(File(gradleProjectDirectory))
	}

	/** Retrieve all files in the `tia/reports` folder sorted by name.  */
	@JvmStatic
	@Throws(IOException::class)
	fun getReportFileNames(mavenProjectPath: String, folderName: String): List<Path> {
		Files.walk(Paths.get(mavenProjectPath, "target", folderName, "reports")).use { stream ->
			return stream.asSequence().filter { Files.isRegularFile(it) }.sorted().toList()
		}
	}

	/** Instructs the agent via HTTP to dump the currently collected coverage.  */
	@JvmStatic
	@Throws(IOException::class)
	fun dumpCoverage(agentPort: Int) {
		Retrofit.Builder().baseUrl("http://localhost:$agentPort").build()
			.create<AgentService>().dump().execute()
	}

	/** Instructs the agent via HTTP to change the partition.  */
	@JvmStatic
	@Throws(IOException::class)
	fun changePartition(agentPort: Int, newPartition: String) {
		val requestBody = newPartition.toRequestBody("text/plain".toMediaTypeOrNull())
		Retrofit.Builder().baseUrl("http://localhost:$agentPort").build()
			.create<AgentService>()
			.changePartition(requestBody)
			.execute()
	}

	/**
	 * Interface for interacting with an agent service that enables coverage dumping and partition management.
	 */
	interface AgentService {
		/** Dumps coverage  */
		@POST("/dump")
		fun dump(): Call<Void>

		/** Changes the current partition.  */
		@PUT("/partition")
		fun changePartition(@Body newPartition: RequestBody): Call<Void>
	}
}
